app/controllers/HasNiphlController.scala

```diff
@@ -82,27 +83,30 @@ class HasNiphlController @Inject() (
             updatedAnswers <-
               Future.fromTry(request.userAnswers.set(HasNiphlUpdatePage, traderProfile.niphlNumber.isDefined))
             _              <- sessionRepository.set(updatedAnswers)
-          } yield Ok(view(form.fill(traderProfile.niphlNumber.isDefined), routes.HasNiphlController.onSubmitUpdate))
-        case Some(value) => Future.successful(Ok(view(form.fill(value), routes.HasNiphlController.onSubmitUpdate)))
+          } yield Ok(
+            view(form.fill(traderProfile.niphlNumber.isDefined), routes.HasNiphlController.onSubmitUpdate(mode))
+          )
+        case Some(value) =>
+          Future.successful(Ok(view(form.fill(value), routes.HasNiphlController.onSubmitUpdate(mode))))
       }
     }
 
-  def onSubmitUpdate: Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
-    form
-      .bindFromRequest()
-      .fold(
-        formWithErrors => Future.successful(BadRequest(view(formWithErrors, routes.HasNiphlController.onSubmitUpdate))),
-        value =>
-          traderProfileConnector.getTraderProfile(request.eori).flatMap { traderProfile =>
-            if (traderProfile.niphlNumber.isDefined == value) {
-              Future.successful(Redirect(routes.ProfileController.onPageLoad()))
-            } else {
+  def onSubmitUpdate(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async {
+    implicit request =>
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors =>
+            Future.successful(BadRequest(view(formWithErrors, routes.HasNiphlController.onSubmitUpdate(mode)))),
+          value =>
+            traderProfileConnector.getTraderProfile(request.eori).flatMap { traderProfile =>
               for {
-                updatedAnswers <- Future.fromTry(request.userAnswers.set(HasNiphlUpdatePage, value))
-                _              <- sessionRepository.set(updatedAnswers)
-              } yield Redirect(navigator.nextPage(HasNiphlUpdatePage, NormalMode, updatedAnswers))
+                updatedAnswers                  <- Future.fromTry(request.userAnswers.set(HasNiphlUpdatePage, value))
+                updatedAnswersWithTraderProfile <-
+                  Future.fromTry(updatedAnswers.set(TraderProfileQuery, traderProfile))
+                _                               <- sessionRepository.set(updatedAnswersWithTraderProfile)
+              } yield Redirect(navigator.nextPage(HasNiphlUpdatePage, NormalMode, updatedAnswersWithTraderProfile))
```

Comment:
should this be mode instead of Normal Mode?

Reply:
yep, i've changed it to mode now